xfrd-tcp.c

```diff
@@ -639,6 +772,80 @@ tcp_conn_ready_for_reading(struct xfrd_tcp* tcp)
 	buffer_clear(tcp->packet);
 }
 
+#ifdef HAVE_TLS_1_3
+static int
+conn_write_ssl(struct xfrd_tcp* tcp, SSL* ssl)
+{
+	ssize_t sent;
+
+	if(tcp->total_bytes < sizeof(tcp->msglen)) {
+		uint16_t sendlen = htons(tcp->msglen);
+		// send
+		int request_length = sizeof(tcp->msglen) - tcp->total_bytes;
+		sent = SSL_write(ssl, (const char*)&sendlen + tcp->total_bytes,
+						 request_length);
+		switch(SSL_get_error(ssl,sent)) {
+			case SSL_ERROR_NONE:
+				if(request_length != sent)
+					log_msg(LOG_ERR, "xfrd: incomplete write of tls");
+				break;
+			default:
+				log_msg(LOG_ERR, "xfrd: generic write problem with tls");
+		}
+
+		if(sent == -1) {
+			if(errno == EAGAIN || errno == EINTR) {
+				/* write would block, try later */
+				return 0;
+			} else {
+				return -1;
+			}
+		}
+
+		tcp->total_bytes += sent;
+		if(sent > (ssize_t)sizeof(tcp->msglen))
+			buffer_skip(tcp->packet, sent-sizeof(tcp->msglen));
+		if(tcp->total_bytes < sizeof(tcp->msglen)) {
+			/* incomplete write, resume later */
+			return 0;
+		}
+		assert(tcp->total_bytes >= sizeof(tcp->msglen));
+	}
+
+	assert(tcp->total_bytes < tcp->msglen + sizeof(tcp->msglen));
+
+	int request_length = buffer_remaining(tcp->packet);
+	sent = SSL_write(ssl, buffer_current(tcp->packet), request_length);
+	switch(SSL_get_error(ssl,sent)) {
+		case SSL_ERROR_NONE:
+			if(request_length != sent)
+				log_msg(LOG_ERR, "xfrd: incomplete write of tls");
```

Comment:
It is normal to have incomplete writes because of the packetizing that happens, so it should not be a priority type LOG_ERR to print it during normal operations.  Perhaps debug or high verbosity printout, but it is not an operational issue.

Reply:
Reporting this error is removed from current code, and 0 is returned to try to write again, so resolved.